code/modules/reagents/chemistry/machinery/reagentgrinder.dm

```diff
@@ -64,334 +48,475 @@
 
+/obj/machinery/reagentgrinder/RefreshParts()
+	. = ..()
+
+	for(var/datum/stock_part/part in component_parts)
+		if(istype(part, /datum/stock_part/servo))
+			speed = part.tier
+		else if(istype(part, /datum/stock_part/matter_bin))
+			maximum_weight = WEIGHT_CLASS_GIGANTIC * part.tier
+/**
+ * Inserts, removes or replaces the beaker present
+ * Arguments
+ *
+ * * mob/living/user - the player performing the action
+ * * obj/item/reagent_containers/new_beaker - the new beaker to replace the old, null to do nothing
+ */
 /obj/machinery/reagentgrinder/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
-	if(!user)
-		return FALSE
-	if(beaker)
+	PRIVATE_PROC(TRUE)
+
+	if(!QDELETED(beaker))
 		try_put_in_hand(beaker, user)
-		beaker = null
-	if(new_beaker)
+
+	if(!QDELETED(new_beaker))
+		if(!user.transferItemToLoc(new_beaker, src))
+			return
 		beaker = new_beaker
-	update_appearance()
-	return TRUE
+
+	update_appearance(UPDATE_OVERLAYS)
+
+/**
+ * Transfer this item or items inside if its a bag into the grinder
+ * Arguments
+ *
+ * * mob/user - the player who is inserting these items
+ * * list/obj/item/to_add - list of items to add
+ */
+/obj/machinery/reagentgrinder/proc/load_items(mob/user, list/obj/item/to_add)
+	PRIVATE_PROC(TRUE)
+
+	//surface level checks to filter out items that can be grinded/juice
+	var/list/obj/item/filtered_list = list()
+	for(var/obj/item/weapon as anything in to_add)
+		//what are we trying to grind exactly?
+		if((weapon.item_flags & ABSTRACT) || (weapon.flags_1 & HOLOGRAM_1))
+			continue
+
+		//Nothing would come from grinding or juicing
+		if(!length(weapon.grind_results) && !weapon.juice_typepath)
+			to_chat(user, span_warning("You cannot grind/juice [weapon] into reagents!"))
+			continue
+
+		//Error messages should be in the objects' definitions
+		if(!weapon.grind_requirements(src))
+			continue
+
+		filtered_list += weapon
+	if(!filtered_list.len)
+		return FALSE
+
+	//find total weight of all items already in grinder
+	var/total_weight
+	for(var/obj/item/to_process in src)
+		if((to_process in component_parts) || to_process == beaker)
+			continue
+		total_weight += to_process.w_class
+
+	//Now transfer the items 1 at a time while ensuring we don't go above the maximum allowed weight
+	var/items_transfered = 0
+	for(var/obj/item/weapon as anything in filtered_list)
+		if(weapon.w_class + total_weight > maximum_weight)
+			to_chat(user, span_warning("[weapon] is too big to fit into [src]"))
+			continue
+		weapon.forceMove(src)
+		total_weight += weapon.w_class
+		items_transfered += 1
+		to_chat(user, span_notice("[weapon] was loaded into [src]"))
+
+	return items_transfered
+
+/obj/machinery/reagentgrinder/item_interaction(mob/living/user, obj/item/tool, list/modifiers, is_right_clicking)
+	if(user.combat_mode || (tool.item_flags & ABSTRACT) || (tool.flags_1 & HOLOGRAM_1) || !can_interact(user) || !user.can_perform_action(src, ALLOW_SILICON_REACH))
```

Comment:
will this allow medical borgs to replace beakers at range

Reply:
No. `user.can_perform_action()` checks for proximity